File: Data Structures & Algorithms/palindrome-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def reverse(self, head: ListNode) -> ListNode:
        cur = head
        prev = None
        while cur:
            tmp = cur.next
            cur.next = prev
            prev = cur
            cur = tmp
        
        return prev

    # 1 -> 2 -> 3 -> N
    # N <- 1 <- 2 <- 3

    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = head
        fast = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        slow = self.reverse(slow)

        fast = head
        while slow:
            if slow.val != fast.val:
                return False
            fast = fast.next
            slow = slow.next
        
        return True
        

File: Data Structures & Algorithms/reorder-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slowList = head; // Slow
        ListNode* fastList = head; // Fast

        while(fastList->next && fastList->next->next){
            slowList = slowList->next;
            fastList = fastList->next->next;
        }

        ListNode* prevFast = nullptr;
        ListNode* curFast = slowList->next;
        while(curFast){
            auto temp = curFast->next;
            curFast->next = prevFast;
            prevFast = curFast;
            curFast = temp;
        }
        slowList->next = nullptr;

        // // Merge two list
        ListNode* cur = head;
        while(prevFast){
            auto temp = cur->next;

            cur->next = prevFast;
            prevFast = prevFast->next;
            cur = cur->next;

            cur->next = temp;
            cur = cur->next;
        }

    }
};
